crux-http-server/src/crux/http_server.clj

```diff
@@ -380,14 +368,167 @@
+     (c/read-edn-string-with-readers formatted-param))))
+
+(defn- build-query
+  [{:strs [find where args order-by limit offset full-results]}]
+  (cond-> {}
+    :find (assoc :find (coerce-param find))
+    :where (assoc :where (coerce-param where :wrap))
+    args (assoc :args (coerce-param args))
+    order-by (assoc :order-by (coerce-param args :wrap))
+    limit (assoc :limit (Integer/parseInt limit))
+    offset (assoc :offset (Integer/parseInt offset))
+    full-results (assoc :full-results? true)))
+
+(defn resolve-headers
+  [{:keys [find full-results?] :as query} results]
+  (if (and full-results? (= 1 (count (first results))))
+    (->> results
+         (mapv #(map keys %))
+         (flatten)
+         (into #{}))
+    find))
+
+(defn resolve-rows
+  [{:keys [find full-results?] :as query} results]
+  (if (and full-results? (= 1 (count (first results))))
+    (map first results)
+    (map #(zipmap find %) results)))
+
+(defn link-top-level-entities
+  [crux-node vt tt path rows]
+  (reduce
+   (fn [coll row]
+     (merge
+      coll
+      (reduce-kv
+       (fn [coll _ v]
+         (if (and (c/valid-id? v)
+                  (api/entity (api/db crux-node vt tt) v))
+           (let [query-params (cond-> "?"
+                                vt (str "valid-time=" vt "&")
+                                tt (str "transact-time=" tt))]
+             (assoc coll v (str path "/" v query-params)))
+           coll))
+       {}
+       row)))
+   {}
+   rows))
+
+(defn query->html [links headers rows]
+  [:html
+   {:lang "en"}
+   [:head
+    [:meta {:charset "utf-8"}]
+    [:meta {:http-equiv "X-UA-Compatible" :content "IE=edge,chrome=1"}]
+    [:meta
+     {:name "viewport"
+      :content "width=device-width, initial-scale=1.0, maximum-scale=1.0"}]
+    [:link {:rel "icon" :href "/favicon.ico" :type "image/x-icon"}]]
+   [:body
+    (if (seq rows)
+      [:table
+       [:thead
+        [:tr
+         (for [header headers]
+           [:th header])]]
+       [:tbody
+        (for [row rows]
+          [:tr
+           (for [header headers]
+             (let [cell-value (get row header)]
+               [:td
+                (if-let [href (get links cell-value)]
+                  [:a {:href href}
+                   (str cell-value)]
+                  (str cell-value))]))])]]
+      [:div "No results found"])]])
+
+(defn data-browser-query [^ICruxAPI crux-node request]
+  (let [query-params (:query-params request)
+        query (build-query query-params)
+        valid-time (some-> (get query-params "valid-time")
+                           (instant/read-instant-date))
+        transaction-time (some-> (get query-params "transaction-time")
+                                 (instant/read-instant-date))
+        db (db-for-request crux-node {:valid-time valid-time
+                                      :transact-time transaction-time})
+        results (api/q db query)]
+    {:status (if (some? results) 200 404)
```

Comment:
suspect this `some?` isn't doing what you think it is - `q` will always return at least an empty set, so `some?` will always return true.

that said, I don't think we want to return 404 on no-results either - it's not like 'entity' where the entity requested in the URL is not found - the the query was valid, it just didn't have any results.